### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Functor): definition of Kan extensions

### PR DESCRIPTION
**Pull request by joelriou**
This PR introduces basic definitions for Kan extensions of functors. It prepares for future development of derived functors and a refactor of the file `CategoryTheory.Limits.KanExtension`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+is a right Kan extension of `F` along `L`. -/
+class IsRightKanExtension : Prop where
+  nonempty_isUniversal : Nonempty (RightExtension.mk F' α).IsUniversal
+
+variable [F'.IsRightKanExtension α]
+
+/-- If `(F', α)` is a right Kan extension of `F` along `L`, then `(F', α)` is a terminal object
+in the category `RightExtension L F`. -/
+noncomputable def isUniversalOfIsRightKanExtension : (RightExtension.mk F' α).IsUniversal :=
+    IsRightKanExtension.nonempty_isUniversal.some
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:79]:
```suggestion
  IsRightKanExtension.nonempty_isUniversal.some
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+/-- If `(F', α)` is a right Kan extension of `F` along `L` and `β : L ⋙ G ⟶ F` is
+a natural transformation, this is the induced morphisms `G ⟶ F'`. -/
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:82]:
```suggestion
a natural transformation, this is the induced morphism `G ⟶ F'`. -/
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+is a left Kan extension of `F` along `L`. -/
+class IsLeftKanExtension : Prop where
+  nonempty_isUniversal : Nonempty (LeftExtension.mk F' α).IsUniversal
+
+variable [F'.IsLeftKanExtension α]
+
+/-- If `(F', α)` is a left Kan extension of `F` along `L`, then `(F', α)` is an initial object
+in the category `LeftExtension L F`. -/
+noncomputable def isUniversalOfIsLeftKanExtension : (LeftExtension.mk F' α).IsUniversal :=
+    IsLeftKanExtension.nonempty_isUniversal.some
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:132]:
```suggestion
  IsLeftKanExtension.nonempty_isUniversal.some
```

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+/-- If `(F', α)` is a left Kan extension of `F` along `L` and `β : F ⟶ L ⋙ G` is
+a natural transformation, this is the induced morphisms `F' ⟶ G`. -/
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:135]:
```suggestion
a natural transformation, this is the induced morphism `F' ⟶ G`. -/
```

File: Mathlib/CategoryTheory/Monoidal/Mon_.lean [hunk 5]
```diff
@@ -412,7 +412,7 @@ theorem Mon_tensor_mul_one (M N : Mon_ C) :
     (M.X ⊗ N.X) ◁ ((λ_ (𝟙_ C)).inv ≫ (M.one ⊗ N.one)) ≫
         tensor_μ C (M.X, N.X) (M.X, N.X) ≫ (M.mul ⊗ N.mul) =
       (ρ_ (M.X ⊗ N.X)).hom := by
-  simp only [whiskerLeft_comp_assoc]
+  simp only [MonoidalCategory.whiskerLeft_comp_assoc]
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Monoidal/Mon_.lean:415]: Although this comment is not directly related to this PR, what do you think about renaming `CategoryTheory.whiskerLeft_comp` to `CategoryTheory.NatTrans.whiskerLeft_comp` or `CategoryTheory.Functor.whiskerLeft_comp` in another PR to avoid this kind of fixes?

File: Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+open Category Limits
+
+namespace Functor
+
+variable {C D H : Type*} [Category C] [Category D] [Category H]
+
+/-- Given two functors `L : C ⥤ H` and `F : C ⥤ D`, this is the category of functors
+`F' : H ⥤ D` equipped with a natural transformation `L ⋙ F' ⟶ F`. -/
+abbrev RightExtension (L : C ⥤ H) (F : C ⥤ D) :=
+  CostructuredArrow ((whiskeringLeft C H D).obj L) F
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Functor/KanExtension/Basic.lean:45]: It may be a matter of preference, but we could write `whiskeringLeft ..` instead of `whiskeringLeft C H D` here.

**Review by yuma-mizuno — APPROVED** [review]
Apart from a few style corrections, it looks great to me!

File: Mathlib/CategoryTheory/Monoidal/Mon_.lean (hunk 5 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Monoidal/Mon_.lean:415]: If we do this name change, in principle, for consistency, we should also rename `CategoryTheory.whiskerLeft` to `CategoryTheory.NatTrans.whiskerLeft`, and similarly `isoWhiskerLeft` -> `NatIso.whiskerLeft`, etc? I am not sure we need to do the change, but I would be ok with it.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge
